src/triage/tracking.py

```diff
@@ -0,0 +1,282 @@
+        experiment_class_name (str) The name of the experiment subclass used
+        experiment_kwargs (dict) Any runtime Experiment keyword arguments that should be saved
+        db_engine (sqlalchemy.engine)
+    """
+    # Any experiment kwargs that are types (e.g. MatrixStorageClass) can't
+    # be serialized, so just use the class name if so
+    cleaned_experiment_kwargs = {
+        k: (v.__name__ if isinstance(v, type) else v)
+        for k, v in experiment_kwargs.items()
+    }
+    run = ExperimentRun(
+        start_time=datetime.datetime.now(),
+        git_hash=infer_git_hash(),
+        triage_version=infer_triage_version(),
+        experiment_hash=experiment_hash,
+        last_updated_time=datetime.datetime.now(),
+        current_status=ExperimentRunStatus.started,
+        installed_libraries=infer_installed_libraries(),
+        platform=platform.platform(),
+        os_user=getpass.getuser(),
+        working_directory=os.getcwd(),
+        ec2_instance_type=infer_ec2_instance_type(),
+        log_location=infer_log_location(),
+        experiment_class_name=experiment_class_name,
+        experiment_kwargs=cleaned_experiment_kwargs,
+        matrices_made=0,
+        matrices_errored=0,
+        matrices_skipped=0,
+        models_made=0,
+        models_errored=0,
+        models_skipped=0,
+    )
+    run_id = None
+    with scoped_session(db_engine) as session:
+        session.add(run)
+        session.commit()
+        run_id = run.run_id
+    if not run_id:
+        raise ValueError("Failed to retrieve run_id from saved row")
+    return run_id
+
+
+@contextmanager
+def get_run_obj_for_update(db_engine, run_id):
+    """Yields an ExperimentRun at the given run_id for update
+
+    Will kick the last_update_time timestamp of the row each time.
+
+    Args:
+        db_engine (sqlalchemy.engine)
+        run_id (int) The identifier/primary key of the run
+    """
+    with scoped_session(db_engine) as session:
+        obj = session.query(ExperimentRun).get(run_id)
+        yield obj
+        obj.last_updated_time = datetime.datetime.now()
+        session.merge(obj)
+
+
+def experiment_entrypoint(entrypoint_func):
+    """Decorator to control tracking of an experiment run at the wrapped method
+
+    To update the database, it requires the instance of the wrapped method to have a
+    db_engine and run_id.
+
+    Upon method entry, will update the ExperimentRun row with the wrapped method name.
+    Upon method exit, will update the ExperimentRun row with the status (either failed or completed)
+    """
+    @wraps(entrypoint_func)
+    def with_entrypoint(*args, **kwargs):
+        entrypoint_name = entrypoint_func.__name__
+        with get_run_obj_for_update(args[0].db_engine, args[0].run_id) as run_obj:
+            if not run_obj.start_method:
+                run_obj.start_method = entrypoint_name
+        try:
+            entrypoint_func(*args, **kwargs)
```

Comment:
Does this not (ever) need to return the function result?

    try:
        return entrypoint_func(self, *args, **kwargs)